### PR TITLE
discv4: new package

### PR DESCRIPTION
**Pull request by ryandotsmith**

### CODE REVIEW

File: eth/eth.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"errors"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"
+	"golang.org/x/crypto/sha3"
+)
+
+type (
+	Hash    [32]byte
+	Address [20]byte
```
Comment (lechengfan) [inline eth/eth.go:14]: where is this type used?

Comment (ryandotsmith) [reply eth/eth.go:14]: Not anywhere. I am just playing around with ideas on this one.

File: discv4/kademlia/table.go [hunk 2]
```diff
@@ -17,10 +18,20 @@ const (
 	minLogDistance = addrByteSize + 1 - bucketsCount
 )
 
+func New() *Table {
+	t := new(Table)
+	for i := 0; i < bucketsCount; i++ {
```
Comment (lechengfan) [inline discv4/kademlia/table.go:23]: Ahh this for loop was what I had initially. Did the zero values not work for `List` and `map` here?

File: discv4/kademlia/table.go [hunk 3]
```diff
@@ -17,10 +18,20 @@ const (
...
+			lru:     list.List{},
+			entries: map[eth.Hash]*list.Element{},
+		}
+	}
+	return t
+}
+
 type Table struct {
-	mu       sync.Mutex // protects buckets
-	selfNode *enr.Record
```
Comment (lechengfan) [inline discv4/kademlia/table.go:22]: I feel pretty strongly that `selfNode` should be an "instance var" here. A Kademlia table, by definition and design, is only supposed to work with a single node as the reference point. Changing the API of `Insert` to take in an arbitrary `self` ENR record is exposing the package to misuse.

Comment (ryandotsmith) [reply discv4/kademlia/table.go:22]: That makes sense. I'll set it up that way.

File: discv4/kademlia/table.go (hunk 2 above)
Comment (ryandotsmith) [reply discv4/kademlia/table.go:23]: Ya, we have to initialize the map. The zero value for the list is good. Just the map.

**Comment by ryandotsmith** [comment]
I marked this as ready for review, but at the last moment, I decided it was best to split this PR into several PRs. I'm going to decompose until we just have a discv4 package to review

File: discv4/kademlia/table.go (hunk 2 above)
Comment (lechengfan) [reply discv4/kademlia/table.go:23]: Good to know, thanks for making the change!

**Comment by ryandotsmith** [comment]
squashed commit history

File: discv4/kademlia/table.go (hunk 3 above)
Comment (ryandotsmith) [reply discv4/kademlia/table.go:22]: added!

**Comment by ryandotsmith** [comment]
ok. finally carved out all the auxiliary changes and what is now left in this pr is pretty focused on adding the discv4 functionality.

**Pull request merged**